packages/mui-base/src/useTabsList/useTabsList.ts

```diff
@@ -81,136 +29,163 @@ const moveFocus = (
-        break;
-      case 'End':
-        event.preventDefault();
-        moveFocus(list, null, previousItem);
-        break;
-      default:
-        break;
-    }
-  };
+  const subitemKeys = React.useMemo(() => Array.from(subitems.keys()), [subitems]);
 
-  const createHandleKeyDown =
-    (otherHandlers: Record<string, React.EventHandler<any>>) => (event: React.KeyboardEvent) => {
-      handleKeyDown(event);
-      otherHandlers.onKeyDown?.(event);
-    };
+  const getTabElement = React.useCallback(
+    (tabValue: string | number) => {
+      if (tabValue == null) {
+        return null;
+      }
 
-  const getRootProps = <TOther extends EventHandlers = {}>(
-    otherHandlers: TOther = {} as TOther,
-  ): UseTabsListRootSlotProps<TOther> => {
-    const propsEventHandlers = extractEventHandlers(parameters);
-    const externalEventHandlers = { ...propsEventHandlers, ...otherHandlers };
+      return subitems.get(tabValue)?.ref.current ?? null;
+    },
+    [subitems],
+  );
 
-    const ownEventHandlers = {
-      onKeyDown: createHandleKeyDown(externalEventHandlers),
-    };
+  const isRtl = direction === 'rtl';
 
-    const mergedEventHandlers = {
-      ...externalEventHandlers,
-      ...ownEventHandlers,
-    };
-    return {
-      'aria-label': ariaLabel,
-      'aria-labelledby': ariaLabelledBy,
-      'aria-orientation': orientation === 'vertical' ? 'vertical' : undefined,
-      role: 'tablist',
-      ref: handleRef,
-      ...mergedEventHandlers,
-    };
-  };
+  let listOrientation: UseListParameters<any, any>['orientation'];
+  if (orientation === 'vertical') {
+    listOrientation = 'vertical';
+  } else {
+    listOrientation = isRtl ? 'horizontal-rtl' : 'horizontal-ltr';
+  }
 
-  const processChildren = React.useCallback(() => {
-    const valueToIndex = new Map();
+  const stateReducer: ListReducer<string | number, ListState<string | number>> = React.useCallback(
+    (state, action) => {
+      const newState = listReducer(state, action);
 
-    let childIndex = 0;
+      if (action.type === ListActionTypes.itemsChange) {
+        return {
+          ...newState,
+          highlightedValue: newState.selectedValues[0] ?? null,
+        };
+      }
 
-    const processedChildren = React.Children.map(children, (child) => {
-      if (!React.isValidElement(child)) {
-        return null;
+      if (action.type === ListActionTypes.setState && action.value.selectedValues != null) {
+        return {
+          ...newState,
+          highlightedValue: action.value.selectedValues[0],
+        };
       }
 
-      if (process.env.NODE_ENV !== 'production') {
-        if (isFragment(child)) {
-          console.error(
-            [
-              "MUI: The Tabs component doesn't accept a Fragment as a child.",
-              'Consider providing an array instead.',
-            ].join('\n'),
-          );
-        }
+      if (selectionFollowsFocus && newState.highlightedValue != null) {
```

Comment:
The beauty of having a reducer - features like `selectionFollowsFocus` are easy and elegant to implement